src/content/contribute/writing-a-loader.md

```diff
@@ -48,138 +41,236 @@ resolveLoader: {
 
-``` text
-url(file) -> require("./file")
-url(~module) -> require("module")
+### Common Code
+
+Avoid generating common code in every module the loader processes. Instead, create a runtime file in the loader and generate a `require` to that shared module.
+
+### Absolute Paths
+
+Don't insert absolute paths into the module code as they break hashing when the root for the project is moved. There's a [`stringifyRequest`](https://github.com/webpack/loader-utils#stringifyrequest) method in `loader-utils` which can be used to convert an absolute path to a relative one.
+
+### Peer Dependencies
+
+If the loader you're working on is a simple wrapper around another package, then you should include the package as a `peerDependency`. This approach allows the application's developer to specify the exact version in the `package.json` if desired.
+
+For instance, the `sass-loader` [specifies `node-sass`](https://github.com/webpack-contrib/sass-loader/blob/master/package.json) as peer dependency like so:
+
+``` js
+"peerDependencies": {
+  "node-sass": "^4.0.0"
+}
 ```
 
-### Extract common code
 
-Don't generate much code that is common in every module processed by that loader. Create a (runtime) file in the loader and generate a `require` to that common code.
+## Testing
 
-## Do not embed absolute paths
+So you've written a loader, followed the guidelines above, and have it set up to run locally. What's next? Let's go through a simple unit testing example to ensure our loader is working the way we expect. We'll be using the [Jest] framework to do this so let's first start by installing and saving it as a dev dependency:
+
+``` bash
+npm i --save-dev jest
+```
 
-Don't put absolute paths in to the module code. They break hashing when the root for the project is moved. There is a method [`stringifyRequest` in loader-utils](https://github.com/webpack/loader-utils#stringifyrequest) which converts an absolute path to an relative one.
+Our loader will process `.txt` files and simply replace any instance of `[name]` with the the `name` option given to the loader. Then it will output a valid JavaScript module containing the text as it's default export:
 
-**Example:**
+__src/loader.js__
 
 ``` js
-var loaderUtils = require("loader-utils");
+import { getOptions } from 'loader-utils';
 
-return "var runtime = require(" +
-  loaderUtils.stringifyRequest(this, "!" + require.resolve("module/runtime")) +
-  ");";
+export default function(source) {
+  const options = getOptions(this);
+
+  source.replace(/\[name\]/g, options.name);
+
+  return `export default ${ JSON.stringify(source) }`;
+};
 ```
 
-### Use a library as `peerDependencies` when they wrap it
+With that in place, we can start setting up our tests. We'll start by adding a configuration that utilizes this loader:
 
-For instance, the [sass-loader specifies node-sass as peer dependency](https://github.com/webpack-contrib/sass-loader/blob/master/package.json):
+__test/webpack.config.js__
```

Comment:
### `Structure`

```
|– src
|– test
| |– __expected__
| |– __snapshots__
| |– fixtures
| | |– fixtures.txt
| | |– fixture.js
| |
| |– helpers
| | |– compiler.js // dynamic webpack.config.js + compiler bootstrap code
| | |– compilation.js // stats/stats.toJson() specific helpers for loader testing
| |
| |– loader.test.js
| |– Errors.test.js
```

### `Setup`

> ℹ️   These should be documented somewhere and maybe an official boilerplate `test-loader` repo

**test/helpers/compiler**
```js
import path from 'path';
import webpack from 'webpack';
import MemoryFS from 'memory-fs';

export default function (fixture, config = {}) {
  // Extend the base config (`webpack.config.js`)  here via config param 
  // TODO document the basic extending possibilities
  // e.g config.loader, config.rule, config.plugins, config.target ...
  config = {
    target: config.target || 'web',
    context: path.resolve(__dirname, '../fixtures'),
    entry: `./${fixture}`,
    output: {
      path: path.resolve(__dirname, `../__expected__/${fixture}`),
      filename: 'bundle.js',
    },
    module: {
      rules: [
        config.rule || config.loader 
          ? {
                test: config.loader.test : /\.txt$/,
                use: {
                  loader: config.loader.loader || '../../src',
                  options: config.loader.options : {}
               }
            : {
                test: /\.txt$/,
                use: {
                  loader: '../../src',
                  options: {}
               }
        }
      ]
    },
    plugins: config.plugins || config.plugin ? [ config.plugin ] : []
  };

  const compiler = webpack(config);
  
 // Keep the output in memory for snapshot testing, otherwise
 // set config.emit = true to write files to `test/__expected__` on disk
  if (!config.emit) {
    compiler.outputFileSystem = new MemoryFS();
  }
 // Run webpack and return the stats for comparison
  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) reject(err);

      resolve(stats);
    });
  });
}
```

**helpers/compilation**
```js
// Needs Triage if helpful/needed (personally not really convinced)
// Loader related stats/stats.toJson() helpers
// e.g

export function loader (stats, i = 1) {
    const modules = stats.toJson().modules
    const { source, map , meta, errors } = modules[i]

    return {
      map,
      meta,
      errors,
      source
    }
}

export function assets (stats) {}
export function modules (stats) {}
export function warnings (stats) {}
...
```

> ℹ️  Folks should start here :)
 
**fixtures/fixture.txt**
```
Hello World
```

**fixtures/fixture.js** (Entry)
```
import txt from './fixture.txt'

export default txt
```

**loader.test.js**
```js
import webpack from './helpers/compiler'

describe('Loader', () => {
   test('Default', async () => {
      const config = {
        loader: {}
      }
      // Run webpack
      const stats = await webpack('fixture.js', config)
      // stats.toJson() and grab what you want to test
      // TODO briefly document what is likely needed here for loaders
      // and links to the Stats Docs
      const result = stats.toJson().modules[1].source // result = 'export default 'HELLO WORLD'

      // for e.g
      expect(result).toMatchSnapshot() 
      // test/__snapshots/loader.test.js.snap
      // exports['Loader Default 1'] = "export default 'HELLO WORLD'"
   })
})
```

### `Examples`

[Simple (`multi-loader`)](https://github.com/webpack-contrib/multi-loader/tree/d3viant0ne-WebpackDefaults/test)
[Advanced (`postcss-loader`)](https://github.com/postcss/postcss-loader/tree/master/test)

Reply:
Hmmm I'll dig into this but I do think we should keep the example in the guide as simple as possible. Obviously linking to a boilerplate repo with examples that vary in difficulty and mentioning a few techniques wouldn't be bad though. 

I'm going to try to set the one up I have locally and if that runs well I'll update the PR and get it merged but you'd be welcome to submit a follow up to add links to these more advanced examples.

Reply:
Yep. It is definitely to complicated as is, but I don't know how to describe the `helpers/compiler` better/briefer atm. Mentioning the structure/idea and slim the file down for docs should be possible though. This helper is fairly important to make testing as 'easy' as it can be.

**helpers/compiler**
```js
import path from 'path';
import webpack from 'webpack';
import MemoryFS from 'memory-fs';

export default function (fixture, config = {}) {
  config = Object.assign({
    context: path.resolve(__dirname, '../fixtures'),
    entry: `./${fixture}`,
    output: {
      path: path.resolve(__dirname, `../__expected__/${fixture}`),
      filename: 'bundle.js',
    }
  }, config);

  const compiler = webpack(config);
  
  if (!config.emit) compiler.outputFileSystem = new MemoryFS();

  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) reject(err);

      resolve(stats);
    });
  });
}
```

Reply:
Ah after a little testing I understand the need for `helpers/compiler`. I was thinking about using `library` in the config to make the `built.js` emit something but using `stats` is cleaner I think and `memory-fs` removes the need for any `.gitignore` or cleanup. Thanks!

Reply:
Yep, exactly 😛